aptoide-installer/src/main/java/cm/aptoide/pt/installer/AptoideInstallPackageInfoMapper.kt

```diff
@@ -2,67 +2,106 @@ package cm.aptoide.pt.installer
-              name = fileName,
-              type = InstallationFile.Type.BASE,
-              md5 = md5,
-              fileSize = filesize,
-              url = path,
-              altUrl = path_alt,
-              localPath = Environment.getExternalStorageDirectory().absolutePath + "/.aptoide/"
-            )
-          )
+        val appMeta = appMetaUseCase.getMetaInfo(app.packageName)
+
+        add(appMeta.file.toInstallationFile(InstallationFile.Type.BASE))
+
+        appMeta.obb?.run {
+          add(main.toInstallationFile(InstallationFile.Type.OBB_MAIN))
+          patch?.let { add(it.toInstallationFile(InstallationFile.Type.OBB_PATCH)) }
         }
-        app.obb?.run {
-          with(main) {
-            add(
-              InstallationFile(
-                name = fileName,
-                type = InstallationFile.Type.OBB_MAIN,
-                md5 = md5,
-                fileSize = filesize,
-                url = path,
-                altUrl = path_alt,
-                localPath = Environment.getExternalStorageDirectory().absolutePath + "/.aptoide/"
-              )
-            )
+
+        appMeta.aab?.run {
+          requiredSplitTypes.forEach { type ->
+            if (splits.find { it.type == type } == null) {
+              throw IllegalStateException("AAB required split types not found")
+            }
           }
-          patch?.run {
-            add(
-              InstallationFile(
-                name = fileName,
-                type = OBB_PATCH,
-                md5 = md5,
-                fileSize = filesize,
-                url = path,
-                altUrl = path_alt,
-                localPath = Environment.getExternalStorageDirectory().absolutePath + "/.aptoide/"
-              )
-            )
+          splits.forEach {
+            add(it.file.toInstallationFile(InstallationFile.Type.PFD_INSTALL_TIME))
+          }
+
+          val dynamicSplits = dynamicSplitsUseCase.getDynamicSplits(app.md5)
+
+          dynamicSplits.forEach {
+            val type = mapDynamicSplitType(it.type, it.deliveryTypes)
+            add(it.file.toInstallationFile(type))
+            addAll(it.splits.map { it.toInstallationFile(type) })
           }
         }
       },
     payload = null
+  ).filter()
+}
+
+private fun File.toInstallationFile(type: InstallationFile.Type) = InstallationFile(
+  name = fileName,
+  type = type,
+  md5 = md5,
+  fileSize = filesize,
+  url = path,
+  altUrl = path_alt,
+  localPath = Environment.getExternalStorageDirectory().absolutePath + "/.aptoide/"
+)
+
+private fun mapDynamicSplitType(
+  splitType: String,
+  deliveryTypes: List<String>,
+) = deliveryTypes.toMutableSet()
+  .let {
+    if (it.remove("INSTALL_TIME")) {
+      "INSTALL_TIME"
+    } else {
+      it.first()
+    }
+  }
+  .let {
+    when (splitType) {
+      "FEATURE" -> "PFD_$it"
+      "ASSET" -> "PAD_$it"
+      else -> throw IllegalArgumentException("Unknown split type")
+    }
+  }
+  .let {
```

Comment:
NIT: You can convert it to the method reference here :-)

Reply:
Fixed